Userland/Applications/Dictionary/Dictionary.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright (c) 2021, Robin Allen <r@foon.uk>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <fcntl.h>
+#include <stdio.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include <AK/StringBuilder.h>
+
+#include "Dictionary.h"
+
+namespace Dictionary {
+
+static char const* part_of_speech_name(char pos)
+{
+    switch (pos) {
+    case 'n':
+        return "noun";
+    case 'v':
+        return "verb";
+    case 's':
+    case 'a':
+        return "adjective";
+    case 'r':
+        return "adverb";
+    default:
+        return "misc";
+    }
+}
+
+static int size_of_index_entry(char const* index_entry)
+{
+    int n = 0;
+    for (int i = 0; i < 16; i++) {
+        if (index_entry[i])
+            n++;
+        else
+            break;
+    }
+    return n;
+}
+
+Dictionary::Dictionary(char const* filename)
+{
+    auto file_or_error = MappedFile::map(filename);
+    if (file_or_error.is_error()) {
+        m_num_words = 0;
+        m_dictionary_bytes = 0;
+        return;
```

Comment:
Is the Dictionary a useful object if it has no data? Returning from the constructor with no words and no bytes is a bit of a :/ moment. I would recommend changing the constructor to take a .. `NonnullRefPtr<MappedFile>`. Then you can make the constructor private and use a static factory function to create one. Either returning an `OwnPtr<Dictionary>`  or a `Result<Dictionary, OsError>` to pass the error from the file mapping creation up to the caller.

You could also make m_dictionary_file a `NonnullRefPtr<MappedFile>` this way.
```c++
Dictionary::Dictionary(NonnullRefPtr<MappedFile> dictionary_file)
   : m_dictionary_file(move(dictionary_file))
{
```

Adopting this design pattern would let your UI layer have some logic to like, throw up a message box and say
> Unable to open default dictionary file "/res/dictionaries/wordnet.bin". Please select a valid dictionary or exit

And throw up a file picker dialog or exit. For example.

It would also let you move the verification of the header validity to the factory function, and let you complain to the user that their database is malformed rather than VERIFY and crash the app.

This pattern would work best if the DictionaryModel was given a reference (or ownership of) a Dictionary object in its constructor, and you did that "dictionary selection" logic in main.cpp.

Reply:
The app already does show a similar message if you haven't built the data file. It determines whether or not to do this by seeing whether or not the dictionary is empty.

About verification of header validity, there is none. At this stage I'm thinking of the data file really as part of the app; you build it along with the app, and it maps it at startup. Like, it might as well be in the executable's `.data` section, it's that tightly-coupled.

I think doing proper validation is super great and necessary once we have user-selectable dictionaries and we're loading them from files associated with the app, but for what we have now I honestly think it's okay.